docs/src/literate-tutorials/gpu_qp_heat_equation.jl

```diff
@@ -0,0 +1,254 @@
+            for j in 1:n_basefuncs
+                ∇u = shape_gradient(qv, j)
+                ## Add contribution to Ke
+                Ke[i, j] += (∇δu ⋅ ∇u) * dΩ
+            end
+        end
+    end
+    return Ke, fe
+end
+
+
+function create_buffers(cellvalues, dh)
+    f = zeros(ndofs(dh))
+    K = create_sparsity_pattern(dh)
+    assembler = start_assemble(K, f)
+    ## Local quantities
+    n_basefuncs = getnbasefunctions(cellvalues)
+    Ke = zeros(n_basefuncs, n_basefuncs)
+    fe = zeros(n_basefuncs)
+    return (;f, K, assembler, Ke, fe)
+end
+
+
+# Standard global assembly
+function assemble_global!(cellvalues, dh::DofHandler,qp_iter::Val{QPiter}) where {QPiter}
+    (;f, K, assembler, Ke, fe) = create_buffers(cellvalues,dh)
+    # Loop over all cels
+    for cell in CellIterator(dh)
+        fill!(Ke, 0)
+        fill!(fe, 0)
+        if QPiter
+            reinit!(cellvalues, getcoordinates(cell))
+            assemble_element_qpiter!(Ke, fe, cellvalues)
+        else
+            # Reinitialize cellvalues for this cell
+            reinit!(cellvalues, cell)
+            # Compute element contribution
+            assemble_element_std!(Ke, fe, cellvalues)
+        end
+        # Assemble Ke and fe into K and f
+        assemble!(assembler, celldofs(cell), Ke, fe)
+    end
+    return K, f
+end
+
+
+
+# Helper function to get all the coordinates from the grid.
+function get_all_coordinates(grid::Ferrite.AbstractGrid{dim}) where {dim}
+    coords = Vector{Vec{2,Float32}}() 
+    n_cells = length(grid.cells)
+    for i = 1:n_cells
+        append!(coords,getcoordinates(grid,i))
+    end
+    coords
+end
+
+struct GPUGrid{sdim,V<:Vec{sdim,Float32},COORDS<:AbstractArray{V,1}} <: Ferrite.AbstractGrid{sdim} 
+    all_coords::COORDS
+    n_cells::Int32
+    
+end
+
+function GPUGrid(grid::Grid{sdim}) where sdim
+    all_coords = cu(get_all_coordinates(grid))
+    n_cells = Int32(length(grid.cells))
+    GPUGrid(all_coords,n_cells)
+end
+
+
+struct GPUDofHandler{CDOFS<:AbstractArray{<:Number,1},GRID<:GPUGrid}<: Ferrite.AbstractDofHandler
+    cell_dofs::CDOFS
+    grid::GRID
+end
+
+
+function GPUDofHandler(dh::DofHandler)
+    GPUDofHandler(cu(Int32.(dh.cell_dofs)),GPUGrid(dh.grid))
+end
+
+function Adapt.adapt_structure(to, grid::GPUGrid)
+    all_coords = Adapt.adapt_structure(to, grid.all_coords)
+    n_cells = Adapt.adapt_structure(to, grid.n_cells)
+    GPUGrid(all_coords, n_cells)
+end
+
+function Adapt.adapt_structure(to, dh::GPUDofHandler)
+    cell_dofs = Adapt.adapt_structure(to, dh.cell_dofs)
+    grid = Adapt.adapt_structure(to, dh.grid)
+    GPUDofHandler(cell_dofs, grid)
+end
+
+function Adapt.adapt_structure(to, cv::Ferrite.StaticCellValues)
+    fv = Adapt.adapt_structure(to, cv.fv)
+    gm = Adapt.adapt_structure(to, cv.gm)
+    x = Adapt.adapt_structure(to, cu(cv.x))
+    weights = Adapt.adapt_structure(to, cv.weights)
+    Ferrite.StaticCellValues(fv,gm,x, weights)
+end
+
+
+gm = static_cellvalues.gm
```

Comment:
Maybe this is still a work in progress but the mix of functions and global variables make it kind of confusing to read the code.